labelbox/pagination.py

```diff
@@ -25,45 +32,90 @@ def __init__(self, client, query, params, dereferencing, obj_class):
                 reach the paginated objects of interest.
             obj_class (type): The class of object to be instantiated with each
                 dict containing db values.
+            cursor_path: If not None, this is used to find the cursor
+            experimental: Used to call experimental endpoints
         """
         self.client = client
         self.query = query
         self.params = params
         self.dereferencing = dereferencing
         self.obj_class = obj_class
+        self.beta = beta
 
-        self._fetched_pages = 0
         self._fetched_all = False
         self._data = []
         self._data_ind = 0
 
+        if cursor_path:
+            self.paginator = _CursorPagination(client, cursor_path)
+        else:
+            self.paginator = _OffsetPagination(client)
+
     def __iter__(self):
         self._data_ind = 0
         return self
 
+    def get_page_data(self, results):
+        for deref in self.dereferencing:
+            results = results[deref]
+        return [self.obj_class(self.client, result) for result in results]
+
     def __next__(self):
         if len(self._data) <= self._data_ind:
             if self._fetched_all:
                 raise StopIteration()
 
-            query = self.query % (self._fetched_pages * _PAGE_SIZE, _PAGE_SIZE)
-            self._fetched_pages += 1
-
-            results = self.client.execute(query, self.params)
-            for deref in self.dereferencing:
-                results = results[deref]
-
-            page_data = [
-                self.obj_class(self.client, result) for result in results
-            ]
+            results = self.paginator.fetch_results(self.query, self.params,
+                                                   self.beta)
+            page_data = self.get_page_data(results)
             self._data.extend(page_data)
+            n_items = len(page_data)
 
-            if len(page_data) < _PAGE_SIZE:
-                self._fetched_all = True
-
-            if len(page_data) == 0:
+            if n_items == 0:
                 raise StopIteration()
 
+            self._fetched_all = self.paginator.fetched_all(n_items, results)
+
         rval = self._data[self._data_ind]
         self._data_ind += 1
         return rval
+
+
+class _CursorPagination:
+
+    def __init__(self, client, cursor_path):
+        self.client = client
+        self.cursor_path = cursor_path
+        self.next_cursor = None
+
+    def get_next_cursor(self, results):
+        for path in self.cursor_path:
+            results = results[path]
+        return results
+
+    def fetched_all(self, n_items, results):
+        self.next_cursor = self.get_next_cursor(results)
+        if self.next_cursor is None:
```

Comment:
You can just do `return self.next_cursor is None`

Reply:
Fixed.